File: src/main/resources/static/js/services/WebSocketSessionRegistry.js
/**
 * WebSocket Session Registry
 * Manages user sessions and subscriptions for the WebSocket server
 */
class WebSocketSessionRegistry {
    constructor() {
        // Maps to track subscriptions by different criteria
        this.sessionMap = new Map(); // sessionId -> Session object
        this.guildSubscriptions = new Map(); // guildId -> Set of sessionIds
        this.channelSubscriptions = new Map(); // channelId -> Set of sessionIds
        this.dmSubscriptions = new Map(); // userId -> Set of sessionIds
        this.userSessions = new Map(); // userId -> Set of sessionIds

        // Session cleanup interval (check every 5 minutes)
        this.cleanupInterval = setInterval(() => this.cleanupSessions(), 5 * 60 * 1000);
    }

    /**
     * Register a new session
     * @param {string} sessionId - WebSocket session ID
     * @param {string} userId - User ID
     * @param {string} username - Username
     * @returns {object} - Session object
     */
    registerSession(sessionId, userId, username) {
        const session = {
            id: sessionId,
            userId: userId,
            username: username,
            createdAt: Date.now(),
            lastActivityAt: Date.now(),
            subscriptions: {
                guilds: new Set(),
                channels: new Set(),
                dms: new Set()
            }
        };

        this.sessionMap.set(sessionId, session);

        // Add to user sessions
        if (!this.userSessions.has(userId)) {
            this.userSessions.set(userId, new Set());
        }
        this.userSessions.get(userId).add(sessionId);

        return session;
    }

    /**
     * Get a session by ID
     * @param {string} sessionId - Session ID
     * @returns {object|null} - Session object or null if not found
     */
    getSession(sessionId) {
        return this.sessionMap.get(sessionId) || null;
    }

    /**
     * Update session activity timestamp
     * @param {string} sessionId - Session ID
     */
    updateSessionActivity(sessionId) {
        const session = this.sessionMap.get(sessionId);
        if (session) {
            session.lastActivityAt = Date.now();
        }
    }

    /**
     * Remove a session
     * @param {string} sessionId - Session ID
     */
    removeSession(sessionId) {
        const session = this.sessionMap.get(sessionId);
        if (!session) return;

        // Remove from user sessions
        const userId = session.userId;
        if (this.userSessions.has(userId)) {
            this.userSessions.get(userId).delete(sessionId);
            if (this.userSessions.get(userId).size === 0) {
                this.userSessions.delete(userId);
            }
        }

        // Remove from guild subscriptions
        session.subscriptions.guilds.forEach(guildId => {
            if (this.guildSubscriptions.has(guildId)) {
                this.guildSubscriptions.get(guildId).delete(sessionId);
                if (this.guildSubscriptions.get(guildId).size === 0) {
                    this.guildSubscriptions.delete(guildId);
                }
            }
        });

        // Remove from channel subscriptions
        session.subscriptions.channels.forEach(channelId => {
            if (this.channelSubscriptions.has(channelId)) {
                this.channelSubscriptions.get(channelId).delete(sessionId);
                if (this.channelSubscriptions.get(channelId).size === 0) {
                    this.channelSubscriptions.delete(channelId);
                }
            }
        });

        // Remove from DM subscriptions
        session.subscriptions.dms.forEach(userId => {
            if (this.dmSubscriptions.has(userId)) {
                this.dmSubscriptions.get(userId).delete(sessionId);
                if (this.dmSubscriptions.get(userId).size === 0) {
                    this.dmSubscriptions.delete(userId);
                }
            }
        });

        // Remove the session itself
        this.sessionMap.delete(sessionId);
    }

    /**
     * Add guild subscription
     * @param {string} guildId - Guild ID
     * @param {string} sessionId - Session ID
     */
    addGuildSubscription(guildId, sessionId) {
        const session = this.sessionMap.get(sessionId);
        if (!session) return;

        // Add to session's guild subscriptions
        session.subscriptions.guilds.add(guildId);

        // Add to guild subscriptions map
        if (!this.guildSubscriptions.has(guildId)) {
            this.guildSubscriptions.set(guildId, new Set());
        }
        this.guildSubscriptions.get(guildId).add(sessionId);

        // Update activity
        this.updateSessionActivity(sessionId);
    }

    /**
     * Add channel subscription
     * @param {string} channelId - Channel ID
     * @param {string} sessionId - Session ID
     */
    addChannelSubscription(channelId, sessionId) {
        const session = this.sessionMap.get(sessionId);
        if (!session) return;

        // Add to session's channel subscriptions
        session.subscriptions.channels.add(channelId);

        // Add to channel subscriptions map
        if (!this.channelSubscriptions.has(channelId)) {
            this.channelSubscriptions.set(channelId, new Set());
        }
        this.channelSubscriptions.get(channelId).add(sessionId);

        // Update activity
        this.updateSessionActivity(sessionId);
    }

    /**
     * Add DM subscription
     * @param {string} userId - User ID to subscribe to DMs with
     * @param {string} sessionId - Session ID
     */
    addDmSubscription(userId, sessionId) {
        const session = this.sessionMap.get(sessionId);
        if (!session) return;

        // Add to session's DM subscriptions
        session.subscriptions.dms.add(userId);

        // Add to DM subscriptions map
        if (!this.dmSubscriptions.has(userId)) {
            this.dmSubscriptions.set(userId, new Set());
        }
        this.dmSubscriptions.get(userId).add(sessionId);

        // Update activity
        this.updateSessionActivity(sessionId);
    }

    /**
     * Remove guild subscription
     * @param {string} guildId - Guild ID
     * @param {string} sessionId - Session ID
     */
    removeGuildSubscription(guildId, sessionId) {
        const session = this.sessionMap.get(sessionId);
        if (!session) return;

        // Remove from session's guild subscriptions
        session.subscriptions.guilds.delete(guildId);

        // Remove from guild subscriptions map
        if (this.guildSubscriptions.has(guildId)) {
            this.guildSubscriptions.get(guildId).delete(sessionId);
            if (this.guildSubscriptions.get(guildId).size === 0) {
                this.guildSubscriptions.delete(guildId);
            }
        }
    }

    /**
     * Remove channel subscription
     * @param {string} channelId - Channel ID
     * @param {string} sessionId - Session ID
     */
    removeChannelSubscription(channelId, sessionId) {
        const session = this.sessionMap.get(sessionId);
        if (!session) return;

        // Remove from session's channel subscriptions
        session.subscriptions.channels.delete(channelId);

        // Remove from channel subscriptions map
        if (this.channelSubscriptions.has(channelId)) {
            this.channelSubscriptions.get(channelId).delete(sessionId);
            if (this.channelSubscriptions.get(channelId).size === 0) {
                this.channelSubscriptions.delete(channelId);
            }
        }
    }

    /**
     * Remove DM subscription
     * @param {string} userId - User ID
     * @param {string} sessionId - Session ID
     */
    removeDmSubscription(userId, sessionId) {
        const session = this.sessionMap.get(sessionId);
        if (!session) return;

        // Remove from session's DM subscriptions
        session.subscriptions.dms.delete(userId);

        // Remove from DM subscriptions map
        if (this.dmSubscriptions.has(userId)) {
            this.dmSubscriptions.get(userId).delete(sessionId);
            if (this.dmSubscriptions.get(userId).size === 0) {
                this.dmSubscriptions.delete(userId);
            }
        }
    }

    /**
     * Get all sessions subscribed to a guild
     * @param {string} guildId - Guild ID
     * @returns {Array} - Array of session objects
     */
    getGuildSubscribers(guildId) {
        if (!this.guildSubscriptions.has(guildId)) {
            return [];
        }

        return Array.from(this.guildSubscriptions.get(guildId))
            .map(sessionId => this.sessionMap.get(sessionId))
            .filter(Boolean); // Filter out any null sessions
    }

    /**
     * Get all sessions subscribed to a channel
     * @param {string} channelId - Channel ID
     * @returns {Array} - Array of session objects
     */
    getChannelSubscribers(channelId) {
        if (!this.channelSubscriptions.has(channelId)) {
            return [];
        }

        return Array.from(this.channelSubscriptions.get(channelId))
            .map(sessionId => this.sessionMap.get(sessionId))
            .filter(Boolean); // Filter out any null sessions
    }

    /**
     * Get all sessions subscribed to DMs with a user
     * @param {string} userId - User ID
     * @returns {Array} - Array of session objects
     */
    getDmSubscribers(userId) {
        if (!this.dmSubscriptions.has(userId)) {
            return [];
        }

        return Array.from(this.dmSubscriptions.get(userId))
            .map(sessionId => this.sessionMap.get(sessionId))
            .filter(Boolean); // Filter out any null sessions
    }

    /**
     * Get all sessions for a user
     * @param {string} userId - User ID
     * @returns {Array} - Array of session objects
     */
    getUserSessions(userId) {
        if (!this.userSessions.has(userId)) {
            return [];
        }

        return Array.from(this.userSessions.get(userId))
            .map(sessionId => this.sessionMap.get(sessionId))
            .filter(Boolean); // Filter out any null sessions
    }

    /**
     * Check if a user is online (has active sessions)
     * @param {string} userId - User ID
     * @returns {boolean} - True if the user has any active sessions
     */
    isUserOnline(userId) {
        return this.userSessions.has(userId) && this.userSessions.get(userId).size > 0;
    }

    /**
     * Get count of active sessions
     * @returns {number} - Number of active sessions
     */
    getSessionCount() {
        return this.sessionMap.size;
    }

    /**
     * Clean up inactive sessions (over 24 hours old)
     */
    cleanupSessions() {
        const now = Date.now();
        const inactivityThreshold = 24 * 60 * 60 * 1000; // 24 hours

        for (const [sessionId, session] of this.sessionMap.entries()) {
            const inactiveTime = now - session.lastActivityAt;

            if (inactiveTime > inactivityThreshold) {
                console.log(`Removing inactive session: ${sessionId}, inactive for ${inactiveTime / 1000} seconds`);
                this.removeSession(sessionId);
            }
        }
    }

    /**
     * Dispose registry and clean up resources
     */
    dispose() {
        if (this.cleanupInterval) {
            clearInterval(this.cleanupInterval);
            this.cleanupInterval = null;
        }

        // Clear all data
        this.sessionMap.clear();
        this.guildSubscriptions.clear();
        this.channelSubscriptions.clear();
        this.dmSubscriptions.clear();
        this.userSessions.clear();
    }
}

// Create global instance
window.sessionRegistry = new WebSocketSessionRegistry();

File: src/main/java/com/cottonlesergal/ubot/websocket/WebSocketSessionRegistry.java
package com.cottonlesergal.ubot.websocket;

import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.messaging.SessionDisconnectEvent;
import org.springframework.context.event.EventListener;

import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArraySet;

/**
 * Registry for WebSocket sessions and subscriptions.
 * Manages tracking which sessions are subscribed to which resources.
 */
@Component
@Slf4j
public class WebSocketSessionRegistry {

    // Maps to track subscriptions
    private final Map<String, Set<String>> channelSubscriptions = new ConcurrentHashMap<>();
    private final Map<String, Set<String>> guildSubscriptions = new ConcurrentHashMap<>();
    private final Map<String, Set<String>> userSubscriptions = new ConcurrentHashMap<>();

    // Reverse maps for cleanup
    private final Map<String, Set<String>> sessionToChannels = new ConcurrentHashMap<>();
    private final Map<String, Set<String>> sessionToGuilds = new ConcurrentHashMap<>();
    private final Map<String, Set<String>> sessionToUsers = new ConcurrentHashMap<>();

    // User session tracking
    private final Map<String, Map<String, Object>> sessionData = new ConcurrentHashMap<>();
    private final Map<String, Set<String>> userToSessions = new ConcurrentHashMap<>();

    /**
     * Register user authentication information with a session
     *
     * @param sessionId The session ID
     * @param userId The user ID
     * @param username The username
     */
    public void registerUserSession(String sessionId, String userId, String username) {
        // Store session data
        Map<String, Object> data = new ConcurrentHashMap<>();
        data.put("userId", userId);
        data.put("username", username);
        data.put("authenticated", true);
        data.put("lastActivity", System.currentTimeMillis());

        sessionData.put(sessionId, data);

        // Map user ID to session for easy lookup
        userToSessions.computeIfAbsent(userId, k -> new CopyOnWriteArraySet<>()).add(sessionId);

        log.debug("Registered user session: {} -> user: {} ({})", sessionId, username, userId);
    }

    /**
     * Check if a session is authenticated
     *
     * @param sessionId The session ID
     * @return true if session is authenticated
     */
    public boolean isSessionAuthenticated(String sessionId) {
        Map<String, Object> data = sessionData.get(sessionId);
        return data != null && Boolean.TRUE.equals(data.get("authenticated"));
    }

    /**
     * Get user ID for a session
     *
     * @param sessionId The session ID
     * @return The user ID or null if not found
     */
    public String getUserIdForSession(String sessionId) {
        Map<String, Object> data = sessionData.get(sessionId);
        return data != null ? (String) data.get("userId") : null;
    }

    /**
     * Get username for a session
     *
     * @param sessionId The session ID
     * @return The username or null if not found
     */
    public String getUsernameForSession(String sessionId) {
        Map<String, Object> data = sessionData.get(sessionId);
        return data != null ? (String) data.get("username") : null;
    }

    /**
     * Get all sessions for a user
     *
     * @param userId The user ID
     * @return Set of session IDs for the user
     */
    public Set<String> getSessionsForUser(String userId) {
        return userToSessions.getOrDefault(userId, Set.of());
    }

    /**
     * Add a channel subscription for a session.
     *
     * @param channelId The channel ID
     * @param sessionId The session ID
     */
    public void addChannelSubscription(String channelId, String sessionId) {
        log.debug("Adding channel subscription: session {} -> channel {}", sessionId, channelId);

        // Add to channel map
        channelSubscriptions.computeIfAbsent(channelId, k -> new CopyOnWriteArraySet<>()).add(sessionId);

        // Add to session map for cleanup
        sessionToChannels.computeIfAbsent(sessionId, k -> new CopyOnWriteArraySet<>()).add(channelId);

        // Update last activity
        updateSessionActivity(sessionId);
    }

    /**
     * Add a guild subscription for a session.
     *
     * @param guildId The guild ID
     * @param sessionId The session ID
     */
    public void addGuildSubscription(String guildId, String sessionId) {
        log.debug("Adding guild subscription: session {} -> guild {}", sessionId, guildId);

        // Add to guild map
        guildSubscriptions.computeIfAbsent(guildId, k -> new CopyOnWriteArraySet<>()).add(sessionId);

        // Add to session map for cleanup
        sessionToGuilds.computeIfAbsent(sessionId, k -> new CopyOnWriteArraySet<>()).add(guildId);

        // Update last activity
        updateSessionActivity(sessionId);
    }

    /**
     * Add a user/DM subscription for a session.
     *
     * @param userId The user ID
     * @param sessionId The session ID
     */
    public void addDmSubscription(String userId, String sessionId) {
        log.debug("Adding user/DM subscription: session {} -> user {}", sessionId, userId);

        // Add to user map
        userSubscriptions.computeIfAbsent(userId, k -> new CopyOnWriteArraySet<>()).add(sessionId);

        // Add to session map for cleanup
        sessionToUsers.computeIfAbsent(sessionId, k -> new CopyOnWriteArraySet<>()).add(userId);

        // Update last activity
        updateSessionActivity(sessionId);
    }

    /**
     * Get sessions subscribed to a channel.
     *
     * @param channelId The channel ID
     * @return Set of session IDs
     */
    public Set<String> getSessionsSubscribedToChannel(String channelId) {
        return channelSubscriptions.getOrDefault(channelId, Set.of());
    }

    /**
     * Get sessions subscribed to a guild.
     *
     * @param guildId The guild ID
     * @return Set of session IDs
     */
    public Set<String> getSessionsSubscribedToGuild(String guildId) {
        return guildSubscriptions.getOrDefault(guildId, Set.of());
    }

    /**
     * Get sessions subscribed to a user.
     *
     * @param userId The user ID
     * @return Set of session IDs
     */
    public Set<String> getSessionsSubscribedToUser(String userId) {
        return userSubscriptions.getOrDefault(userId, Set.of());
    }

    /**
     * Remove all subscriptions for a session.
     *
     * @param sessionId The session ID
     */
    public void removeAllSubscriptionsForSession(String sessionId) {
        log.debug("Removing all subscriptions for session: {}", sessionId);

        // Remove channel subscriptions
        Set<String> channelIds = sessionToChannels.remove(sessionId);
        if (channelIds != null) {
            channelIds.forEach(channelId -> {
                Set<String> sessions = channelSubscriptions.get(channelId);
                if (sessions != null) {
                    sessions.remove(sessionId);
                    if (sessions.isEmpty()) {
                        channelSubscriptions.remove(channelId);
                    }
                }
            });
        }

        // Remove guild subscriptions
        Set<String> guildIds = sessionToGuilds.remove(sessionId);
        if (guildIds != null) {
            guildIds.forEach(guildId -> {
                Set<String> sessions = guildSubscriptions.get(guildId);
                if (sessions != null) {
                    sessions.remove(sessionId);
                    if (sessions.isEmpty()) {
                        guildSubscriptions.remove(guildId);
                    }
                }
            });
        }

        // Remove user subscriptions
        Set<String> userIds = sessionToUsers.remove(sessionId);
        if (userIds != null) {
            userIds.forEach(userId -> {
                Set<String> sessions = userSubscriptions.get(userId);
                if (sessions != null) {
                    sessions.remove(sessionId);
                    if (sessions.isEmpty()) {
                        userSubscriptions.remove(userId);
                    }
                }
            });
        }

        // Clean up session data
        Map<String, Object> data = sessionData.remove(sessionId);
        if (data != null && data.get("userId") != null) {
            String userId = (String) data.get("userId");
            Set<String> userSessions = userToSessions.get(userId);
            if (userSessions != null) {
                userSessions.remove(sessionId);
                if (userSessions.isEmpty()) {
                    userToSessions.remove(userId);
                }
            }
        }
    }

    /**
     * Update session activity timestamp
     *
     * @param sessionId The session ID
     */
    private void updateSessionActivity(String sessionId) {
        Map<String, Object> data = sessionData.get(sessionId);
        if (data != null) {
            data.put("lastActivity", System.currentTimeMillis());
        }
    }

    /**
     * Event listener for session disconnect events.
     * Automatically cleans up subscriptions when a WebSocket session disconnects.
     *
     * @param event The disconnect event
     */
    @EventListener
    public void handleSessionDisconnect(SessionDisconnectEvent event) {
        String sessionId = event.getSessionId();
        log.info("WebSocket session disconnected: {}", sessionId);
        removeAllSubscriptionsForSession(sessionId);
    }

    /**
     * Get all authenticated sessions
     *
     * @return Set of authenticated session IDs
     */
    public Set<String> getAuthenticatedSessions() {
        Set<String> authenticatedSessions = new CopyOnWriteArraySet<>();

        sessionData.forEach((sessionId, data) -> {
            if (Boolean.TRUE.equals(data.get("authenticated"))) {
                authenticatedSessions.add(sessionId);
            }
        });

        return authenticatedSessions;
    }

    /**
     * Get session count
     *
     * @return Total number of active sessions
     */
    public int getSessionCount() {
        return sessionData.size();
    }

    /**
     * Get authenticated session count
     *
     * @return Number of authenticated sessions
     */
    public int getAuthenticatedSessionCount() {
        return (int) sessionData.values().stream()
                .filter(data -> Boolean.TRUE.equals(data.get("authenticated")))
                .count();
    }
}